src/run_pretrain.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2021 Google LLC
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     https://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+
+import tensorflow as tf
+from official.modeling import tf_utils
+
+import modeling
+import input_utils
+import tensor_utils
+
+
+def model_fn_builder(model_config, init_checkpoint, learning_rate,
+                     num_train_steps, num_warmup_steps, use_tpu,
+                     use_one_hot_embeddings, optimizer, poly_power,
+                     start_warmup_step, learning_rate_schedule):
+  """Returns `model_fn` closure for TPUEstimator."""
+
+  def model_fn(features, labels, mode, params):
+    raise NotImplementedError
+
+  return model_fn
+
+
+def metric_fn(lm_example_loss,
+              masked_lm_log_probs,
+              masked_lm_ids,
+              masked_lm_weights,
+              is_train=False):
+  """Computes the loss and accuracy of the model."""
+  raise NotImplementedError
+
+
+def get_mpp_output(model_config, input_tensor, output_num_classes, positions,
+                   label_ids, label_weights):
+
+  """Get loss and log probs for the masked patch prediction (MPP)."""
+  input_tensor = tensor_utils.gather_indexes(input_tensor, positions)
+
+  # NOTE (roylu): Do we need layer norm before the linear projection? 
```

Comment:
Not sure either. Does VIT include this layer norm before computing the logits?

Reply:
I followed the [Pytorch implementation](https://github.com/lucidrains/vit-pytorch/blob/22da26fa4b7d98c2a936b48c1b1dcf83c404dfb8/vit_pytorch/vit.py#L126). Not sure if it's necessary or exists in original implementation.

Reply:
We could probably try both.